src/tqec/compile/specs/library/base.py

```diff
@@ -0,0 +1,485 @@
+        return Block(layers)
+
+
+# TODO: finish the implementation of the BasePipeBuilder
+class BasePipeBuilder(PipeBuilder):
+    """Base implementation of the
+    :class:`~tqec.compile.specs.base.PipeBuilder` interface.
+
+    This class provides a good enough default implementation that should be
+    enough for most of the block builders.
+    """
+
+    def __init__(self, compiler: PlaquetteCompiler) -> None:
+        """Initialise the :class:`BaseSubstitutionBuilder` with a compiler.
+
+        Args:
+            compiler: compiler to transform the plaquettes in the standard
+                implementation to a custom implementation.
+        """
+        self._translator = DefaultRPNGTranslator()
+        self._compiler = compiler
+
+    def _get_plaquette(self, description: RPNGDescription) -> Plaquette:
+        return self._compiler.compile(self._translator.translate(description))
+
+    def __call__(self, spec: PipeSpec) -> Block:
+        if spec.pipe_kind.is_temporal:
+            return self.get_temporal_pipe_block(spec)
+        return self.get_spatial_pipe_block(spec)
+
+    #######################
+    #    TEMPORAL PIPE    #
+    #######################
+
+    def get_temporal_pipe_block(self, spec: PipeSpec) -> Block:
+        """Returns the block to implement a temporal pipe based on the
+        provided ``spec``.
+
+        Args:
+            spec: description of the pipe that should be implemented by this
+                method. Should be a temporal pipe.
+
+        Raises:
+            AssertionError: if ``spec`` does not represent a temporal junction.
+
+        Returns:
+            the block to implement a temporal pipe based on the
+        provided ``spec``..
+        """
+        assert spec.pipe_kind.is_temporal
+        if spec.pipe_kind.has_hadamard:
+            return self._get_temporal_hadamard_pipe_block(spec)
+        # Else, it is a regular temporal junction
+        return self._get_temporal_non_hadamard_pipe_block(spec)
+
+    def _get_temporal_non_hadamard_pipe_block(self, spec: PipeSpec) -> Block:
+        """Returns the block to implement a regular temporal junction
+        without Hadamard transition.
+
+        Args:
+            spec: description of the pipe that should be implemented by this
+                method. Should be a regular (i.e., non-Hadamard) temporal pipe.
+
+        Raises:
+            AssertionError: if the provided ``pipe`` is not a temporal pipe, or
+                if it contains a Hadamard transition.
+
+        Returns:
+            the block to implement the provided
+            ``spec``.
+        """
+        assert spec.pipe_kind.is_temporal
+        assert not spec.pipe_kind.has_hadamard
+
+        z_observable_orientation = (
+            ZObservableOrientation.HORIZONTAL
+            if spec.pipe_kind.x == Basis.Z
+            else ZObservableOrientation.VERTICAL
+        )
+        memory_plaquettes = get_memory_qubit_plaquettes(z_observable_orientation)
```

Comment:
Should rather use the `*_rpng_descriptions` variant along with `_get_plaquettes` to use user-provided translator/compiler.